File: implementations/python/process.py
import json
from functools import reduce
import os


def aggregateBatch(list):
    (sum, count) = reduce(lambda acc, el: (
        acc[0]+el[0]*el[1], acc[1]+el[1]), list, (0, 0))
    return sum/count


def getScale(metric):
    if metric.startswith("s"):
        return 1
    elif metric.startswith("ms"):
        return 1e-3
    elif metric.startswith("us"):
        return 1e-6
    elif metric.startswith("ns"):
        return 1e-9
    else:
        raise ValueError("unhandled metric encountered!")


def processFile(rawFilepath, destinationFilepath):
    with open(rawFilepath, "r") as input, open(destinationFilepath, "w") as output:
        inJson = json.load(input)
        rawData = inJson[0]["primaryMetric"]["rawDataHistogram"]
        metric = inJson[0]["primaryMetric"]["scoreUnit"]
        scale = getScale(metric)
        processed = [[scale*aggregateBatch(batch) for batch in run]
                     for run in rawData]
        json.dump(processed, output, separators=(',', ':'))


def processAllFiles():
    rawPath = "/raw/"
    destPath = "/processed/"

    for file in os.listdir(rawPath):
        print(file)
        processFile(rawPath + file, destPath + file)


def main():
    processAllFiles()


if __name__ == "__main__":
    main()

File: implementations/R_rjson/process.R
library(rjson)
library(purrr)

aggregate_batch <- function(batch) {
    sum_and_count <- Reduce(
        function(acc, el) c(acc[1] + el[1] * el[2], acc[2] + el[2]),
        batch, c(0, 0)
    )

    sum_and_count[1] / sum_and_count[2]
}

get_scale <- function(metric) {
    if (startsWith(metric, "s")) {
        scale <- 1
    } else if (startsWith(metric, "ms")) {
        scale <- 1e-3
    } else if (startsWith(metric, "us")) {
        scale <- 1e-6
    } else if (startsWith(metric, "ns")) {
        scale <- 1e-9
    } else {
        stop("unhandled metric encountered")
    }

    scale
}

process_file <- function(raw_filepath, destination_filepath) {
    in_json <- fromJSON(file = raw_filepath)
    raw_data <- in_json[[1]]$primaryMetric$rawDataHistogram
    metric <- in_json[[1]]$primaryMetric$scoreUnit
    scale <- get_scale(metric)
    processed <- lapply(raw_data, function(run) lapply(run, function(batch) scale*aggregate_batch(batch)))
    out_json <- toJSON(processed)
    write(out_json, destination_filepath)
}

processAllFiles <- function() {
    raw_path <- "/raw/"
    dest_path <- "/processed/"
    dir.create(dest_path, showWarnings = FALSE, recursive = TRUE)

    for (filename in list.files(raw_path)) {
        print(filename)
        process_file(paste(raw_path, filename, sep = ""), paste(dest_path, filename, sep = ""))
    }
}

processAllFiles()
